keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -348,25 +350,164 @@ to fail with a descriptive error.
+- `format`: Since CEL supports string concatenation, the value of having format would only be to do things like format floats.
+  Formatting functions are complex and require extensive documentation to teach.
+- `isSorted` makes it possible to check if a list is sorted without the expense of performing a sort
+- `indexOf` / `lastIndexOf` / `split` / `replace` (the string functions) will be overloaded to provide the equivalent list functions.
+- `reduce` is going to be non-obvious to anyone that doesn't have a functional programming background?
+  - If we provide `reduce` are developers going to also expect to have `fold` or `zip`?
+- "Ability to construct a map using a comprehension": this appears mechanically problematic to support in CEL?
+- None of the other policy libraries provide extended math/trig support and I asked Tristan (maintainer of CEL) if
+  they are requested and he said "almost never", which he clarified to mean it has been requested exactly one time.
+- Average and quantile (median, 99th percentile, ...) and stddev are, however, often requested by other CEL users
+
+Future work:
+
+We've decided NOT to add the following functions. They may be useful for mutation, in which case we will consider adding them as part
+of any future work done involving CEL and mutating admission control:
+
+- Regex replace
+- Add `trimPrefix` / `trimSuffix` for strings
+- Add `trim` / `trimLeft` / `trimRight` (overloaded to take an optional cutset arg) for strings
+- `sublist`, `split`, `replace` and `reverse` for lists (overloading existing string functions where appropriate)
+
+##### Function Library Updates
+
+Any changes to the CEL function library will make it possible for CRD authors to create CRDs that are incompatible with
+all previous Kubernetes releases that supported CEL. Because of this incompatibility, changing the function library
+will need to carefully considered and kept to a minimum. All changes will need to be limited to function additions.
+
+Any function library change must follow the [API Changes](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md)
+guidelines. Since a change to the function library is a change to the `x-kubernetes-validations.rule` field, it must be
+introduced one Kubernetes release prior to when it may be included in create/update requests to CRDs. Specifically:
+
+- Kubernetes Version N-1: Entirely unaware of the library update.
+- Kubernetes Version N: Supports CRD that use library updates, but does not allow the library updates to used when
+  setting `x-kubernetes-validations.rule` fields.
+- Kubernetes Version N+1: Library update fully enabled.
+
+The mechanism for this will be:
+
+- All new functions, macros, or overloads of existing functions, will be added to a separate "kubernetes-future-compatibility" CEL extension library. (Better naming suggestions welcome).
```

Comment:
> Better naming suggestions welcome

if the library name is not exposed via the API, I don't really care what we call it

Reply:
👍  It's just an internal name in the code.